### ISSUE TITLE
improve hs.pasteboard image handling

### ISSUE THREAD

**Issue by heptal** [Hammerspoon/hammerspoon#737]
Right now 'getImageContents' only works for PNG data, which is fine for screenshots directly to clipboard. It seems copying in other contexts loads it onto the pasteboard as "public.tiff" as shown by `hs.inspect(hs.pasteboard.contentTypes())`. TIFF is apparently being used as a generic container for 'image data that is not PNG or not PNG loaded via screenshot-to-pasteboard or a few other techniques such as hs.pasteboard.setImageContents', or, 'pretty much all image data with a few exceptions'. 

See [How do I put a GIF onto an NSPasteboard?](http://www.cocoabuilder.com/archive/cocoa/151166-how-do-put-gif-onto-an-nspasteboard.html)
>The usual way is to use an NSFilesPromisePboardType, which was a longer
discussion in this list a long time ago.
Or you can write the gif data to the filesystem and use the
NSFilenamesPboardType.
Or you cheat and write the gif data to the NSTIFFPboardType

It seems Apple has taken the 'cheating' method instead of creating pasteboard types for all image formats. Today it is called `NSPasteboardTypeTIFF`. 

I intend on tackling this one unless @asmagill has done any work on supporting more pasteboard types as I saw him mention awhile back.

**Comment by asmagill** [comment]
I started playing with NSPasteboard's `readObjectsForClasses:options:` and `writeObjects:`, which allow directly getting and putting Cocoa framework classes which support the NSPasteboardWriting and NSPasteboardReading protocols (NSImage does) onto the pasteboard without a whole lot of hassle, but I haven't gotten to NSImage yet (hoping to tonight or tomorrow)...  Adding support for promised types does not look trivial if we want to make it extendable and not just for a few types we hand-select, but I may be missing something.

I know we can load a GIF image from a file into hs.image, which is at heart just an NSImage object, so I'm hopeful that using `writeObjects:` may make this easier, but (if I understand correctly, which is a big if) this only works if the receiver can take an NSImage and apply it in the way it wants... for testing purposes, do you have a specific use case I can try out (an application you are getting from or sending to)?

I'll push what I have tonight if you want to try it out yourself or extend it... so far, I'm just adding what I'm referring to as "experimental" functions/methods to hs.pasteboard, so it shouldn't break anything you already have in place.

**Comment by heptal** [comment]
GIF Information gets lost going to or from the pasteboard

Save to `~/Pictures/pikachu.gif` or what have you: ![](http://i.imgur.com/jDKtxIN.gif)

```lua
function imagePreview(image)
    local mousepoint = hs.mouse.getAbsolutePosition()
    local imageRect = hs.drawing.image(hs.geometry.rect(mousepoint.x+20, mousepoint.y, 100, 100), image)
    imageRect:show()
    hs.timer.doAfter(3, function() imageRect:delete() end)
end

local gif1=hs.image.imageFromPath("~/Pictures/pikachu.gif")
imagePreview(gif1) --works fine, is animated

hs.pasteboard.writeObjects(gif1)

hs.inspect(hs.pasteboard.contentTypes())
--{ "public.tiff" }

gif2=hs.pasteboard.readImage()
imagePreview(gif2) --not animated, first frame only

```
I think the situation may be hopeless as OSX is so weird about GIFs
Now if I right click and copy image from Safari:
```lua
hs.inspect(hs.pasteboard.allContentTypes())

--Copied from Safari:
{ { "public.tiff", "dyn.ah62d4rv4gu8zs3pcnzme2641rf4guzdmsv0gn64uqm10c6xenv61a3k", "dyn.ah62d4rv4gu8yc6durvwwaznwmuuha2pxsvw0e55bsmwca7d3sbwu", "public.url", "public.url-name", "public.utf8-plain-text", "com.apple.flat-rtfd" } }

--From Finder
{ { "public.file-url", "com.apple.icns", "public.utf16-external-plain-text", "public.utf8-plain-text" } }

--From Messages
{ { "public.url", "public.file-url", "public.tiff" } }

--From Chrome
{ { "public.tiff" } }

--From Preview
{ { "public.tiff", "dyn.ah62d4rv4gu8y6y4grf0gn5xbrzw1gydcr7u1e3cytf2gn", "public.utf8-plain-text", "public.file-url" }, { "public.file-url" }, { "public.file-url" }, { "public.file-url" }, { "public.file-url" }, { "public.file-url" }, { "public.file-url" }, { "public.file-url" }, { "public.file-url" }, { "public.file-url" }, { "public.file-url" }, { "public.file-url" }, { "public.file-url" }, { "public.file-url" }, { "public.file-url" }, { "public.file-url" }, { "public.file-url" } }

```
Now if I paste from each of those sources into Messages, all are animated except the ones copied from Chrome and Preview, which are still frames. If I paste from those sources into Finder, nothing will happen if copied from Chrome or Safari, it will create a proper animated .GIF if copied from Messages or Finder, and if copied from Preview, it will create a bunch of .TIFF files (one for each frame).

TL;DR: .GIF animation is completely b0rked in the OSX pasteboard and we should probably forget about it :)

**Comment by heptal** [comment]
Other than GIFs which are seemingly hopeless, #744 addresses the other goals I had after testing.

**Comment by asmagill** [comment]
I wouldn't say GIF's are hopeless, but they are not trivial... and as you saw, not evenly handled across even Apple's own applications... I guess it comes down to a question of need...

I'll go ahead and make to adjustments to include the additions I've worked on and include a note here that if the need for GIF's is revisited in the future, I currently believe that it will require wrapping the raw GIF data in an NSPasteboardItem object and then including either multiple items on the clipboard (the raw GIF data and an NSImage representation for applications that don't explicitly ask for `com.compuserve.gif`) or promising the GIF version for `com.compuserve.gif` and writing a callback (Objective-C, probably not Lua) to handle the promise.

**Comment by ianbytchek** [comment]
@asmagill @heptal hey guys! I'm fighting same issues while attempting to provide copy support for animated gifs. Did you figure out anything in the end or completely given up on this?

**Comment by asmagill** [comment]
I'm not aware of anyone currently working on this, and I won't have the time to revisit it for awhile.

Hammerspoon is using NSPasteboard's `readObject*` and `writeObject*` methods to handle the conversion into and out of NSImage objects (which is what `hs.image` is a wrapper for) and it appears that this opts for the tiff representation, when available, as the preferred type.  If a file url representation is available, you might have more luck working with that, but I haven't really tried myself yet.
